src/clinical/id-generator.ts

```diff
@@ -0,0 +1,267 @@
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
+ * TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+ * OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
+ * IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+import axios from 'axios';
+import memoize from 'memoizee';
+import ms from 'ms';
+import { Donor } from './clinical-entities';
+import { ClinicalEntitySchemaNames } from '../common-model/entities';
+import { config } from '../config';
+import { Errors } from '../utils';
+
+interface IdGenerationRequest {
+	programId: string;
+	submitterSpecimenId: string;
+	submitterSampleId: string;
+	submitterPrimaryDiagnosisId: string;
+	submitterFollowUpId: string;
+	submitterTreatmentId: string;
+	submitterDonorId: string;
+	testInterval: string;
+	family_relative_id: string;
+	comorbidityTypeCode: string;
+	entityType: string;
+}
+
+const request = {
+	programId: '-',
+	submitterSpecimenId: '-',
+	submitterSampleId: '-',
+	submitterPrimaryDiagnosisId: '-',
+	submitterFollowUpId: '-',
+	submitterTreatmentId: '-',
+	submitterDonorId: '-',
+	testInterval: '-',
+	family_relative_id: '-',
+	comorbidityTypeCode: '-',
+	entityType: '-',
+};
+
+export interface PartialDonor extends Partial<Donor> {}
+
+export async function getId(req: IdGenerationRequest) {
+	console.log('getId function called: ' + req.submitterDonorId);
+	const token = await getToken();
+	const headers = {
+		headers: {
+			Authorization: 'Bearer ' + token,
+		},
+	};
+
+	let response;
+	try {
+		const url =
+			config.getConfig().idServiceUrl() +
+			`${req.programId}/${req.submitterDonorId}/${req.submitterSpecimenId}/${req.submitterSampleId}/${req.submitterPrimaryDiagnosisId}/${req.submitterFollowUpId}/${req.submitterTreatmentId}/${req.testInterval}/${req.family_relative_id}/${req.comorbidityTypeCode}/${req.entityType}`;
+		response = await axios.get(url, headers);
+		console.log('getId response: ' + response.data.entityId + ' - ' + response.data.entityType);
+	} catch (e) {
+		throw new Errors.NetworkError('Error sending request to ID service. Caused by: ' + e);
+	}
+	if (!response.data.entityId) {
+		throw new Errors.IdGenerationError('Error generating entity ids.');
+	}
+	return parseInt(response.data.entityId);
+}
+
+export async function setEntityIdsForDonors(donors: Donor[]) {
+	const donorsWithIds = donors.map(async (donor) => {
+		return await setEntityIds(donor);
+	});
+	return await Promise.all(donorsWithIds);
+}
+
+export async function setEntityIds(donor: PartialDonor) {
+	const submitterDonorId = donor.submitterId as string;
+	// -- DONOR --
+	const donorId = await getId({
+		...request,
+		programId: donor.programId as string,
+		submitterDonorId: donor.submitterId as string,
```

Comment:
We don't know that these are strings.

Reply:
Not repeating this comment, but this same pattern continues where the values may be undefined but are not being checked. This could result in requests to the ID service with `undefined` values in the path params.